swaps/swaps_service.go

```diff
@@ -125,254 +183,1057 @@ func (svc *swapsService) EnableAutoSwaps(ctx context.Context, lnClient lnclient.
+		logger.Logger.WithFields(logrus.Fields{
+			"swapId":        swapId,
+			"refundAddress": address,
+		}).WithError(err).Error("Failed to save refund address to swap")
+		return err
+	}
+
+	refundTransaction, _, err := boltz.ConstructTransaction(
+		network,
+		boltz.CurrencyBtc,
+		[]boltz.OutputDetails{
+			{
+				SwapId:             swapId,
+				SwapType:           boltz.NormalSwap,
+				Address:            address,
+				LockupTransaction:  lockupTransaction,
+				TimeoutBlockHeight: swapTransactionResp.TimeoutBlockHeight,
+				Vout:               vout,
+				PrivateKey:         ourKeys,
+				SwapTree:           tree,
+				Cooperative:        true,
+			},
+		},
+		float64(feeRates.FastestFee),
+		svc.boltzApi,
+	)
+	if err != nil {
+		logger.Logger.WithField("swapId", swapId).WithError(err).Error("Could not create claim transaction")
+		return err
+	}
+
+	vout, _, _ = refundTransaction.FindVout(network, address)
+	refundAmount, _ := refundTransaction.VoutValue(vout)
+
+	txHex, err := refundTransaction.Serialize()
+	if err != nil {
+		logger.Logger.WithField("swapId", swapId).WithError(err).Error("Could not serialize refund transaction")
+		return err
+	}
+
+	// TODO: Replace with LNClient broadcast method to avoid trusting boltz
+	claimTxId, err := svc.boltzApi.BroadcastTransaction(boltz.CurrencyBtc, txHex)
+	if err != nil {
+		logger.Logger.WithField("swapId", swapId).WithError(err).Error("Could not broadcast transaction")
+		return err
+	}
+
+	logger.Logger.WithFields(logrus.Fields{
+		"swapId":    swapId,
+		"claimTxId": claimTxId,
+	}).Info("Claim transaction broadcasted for refund")
+
+	err = svc.db.Model(&swap).Updates(&db.Swap{
+		ClaimTxId:     claimTxId,
+		ReceiveAmount: refundAmount,
+		State:         constants.SWAP_STATE_REFUNDED,
+	}).Error
+	if err != nil {
+		logger.Logger.WithFields(logrus.Fields{
+			"swapId":    swapId,
+			"claimTxId": claimTxId,
+		}).WithError(err).Error("Failed to save claim txid to swap")
+		return err
+	}
+
+	return nil
+}
+
+func (svc *swapsService) GetSwap(swapId string) (*Swap, error) {
+	var swap db.Swap
+	err := svc.db.Limit(1).Find(&swap, &db.Swap{
+		SwapId: swapId,
+	}).Error
+
+	if err != nil {
+		logger.Logger.WithError(err).Error("Failed to get swap")
+		return nil, err
+	}
+
+	return &swap, nil
+}
+
+func (svc *swapsService) ListSwaps() ([]Swap, error) {
+	var swaps []db.Swap
+	err := svc.db.Find(&swaps).Error
+
+	if err != nil {
+		logger.Logger.WithError(err).Error("Failed to list swaps")
+		return nil, err
+	}
+
+	return swaps, nil
+}
+
+func (svc *swapsService) subscribePendingSwaps() {
+	var swaps []db.Swap
+	if err := svc.db.Where("state = ?", constants.SWAP_STATE_PENDING).Find(&swaps).Error; err != nil {
+		logger.Logger.WithError(err).Error("failed to load pending swaps")
+		return
+	}
+	if len(swaps) == 0 {
+		return
+	}
+
+	logger.Logger.WithField("count", len(swaps)).Info("Resuming pending swaps...")
+
+	for _, swap := range swaps {
+		var boltzWs *boltz.Websocket
```

Comment:
We should actually only need one websocket in total shouldn't we?

Reply:
@im-adithya hmm, can you verify that and maybe ask Boltz? that doesn't sound good.

Reply:
Resolved (for pending swaps we now only use a single websocket connection), we can do this similarly for individual swaps too but it's not possible currently because we can't chain swap id subscriptions right now on a single boltz websocket connection. Reached out to them on discord.